src/Processors/Sources/MongoDBSource.cpp

```diff
@@ -1,465 +1,489 @@
+            return json;
+        }
+        case bsoncxx::type::k_binary:
+            return base64Encode(std::string(reinterpret_cast<const char*>(value.get_binary().bytes), value.get_binary().size));
+        case bsoncxx::type::k_undefined:
+            return "undefined";
+        case bsoncxx::type::k_oid:
+            return value.get_oid().value.to_string();
+        case bsoncxx::type::k_bool:
+            return value.get_bool().value ? "true" : "false";
+        case bsoncxx::type::k_date:
+            return DateLUT::instance().timeToString(value.get_date().to_int64() / 1000);
+        case bsoncxx::type::k_null:
+            return "null";
+        case bsoncxx::type::k_regex:
+            return R"({"regex": ")" + escapeJSONString(std::string(value.get_regex().regex)) + R"(","options":")" + escapeJSONString(std::string(value.get_regex().regex)) + "\"}";
+        case bsoncxx::type::k_dbpointer:
+            return "{\"" + escapeJSONString(value.get_dbpointer().value.to_string()) + "\":\"" + escapeJSONString(std::string(value.get_dbpointer().collection)) + "\"}";
+        case bsoncxx::type::k_symbol:
+            return {1, value.get_symbol().symbol.at(0)};
+        case bsoncxx::type::k_int32:
+            return std::to_string(static_cast<Int64>(value.get_int32().value));
+        case bsoncxx::type::k_timestamp:
+            return DateLUT::instance().timeToString(value.get_timestamp().timestamp);
+        case bsoncxx::type::k_int64:
+            return std::to_string(value.get_int64().value);
+        case bsoncxx::type::k_decimal128:
+            return value.get_decimal128().value.to_string();
+        default:
+            throw Exception(ErrorCodes::NOT_IMPLEMENTED, "BSON type {} is unserializable", to_string(value.type()));
+    }
+}
 
-                auto array = static_cast<const Poco::MongoDB::ConcreteElement<MongoArray::Ptr> &>(value).value();
+template <typename T, typename T2>
+T BSONElementAsNumber(const T2 & value, const std::string & name)
+{
+    switch (value.type())
+    {
+        case bsoncxx::type::k_int64:
+            return static_cast<T>(value.get_int64());
+        case bsoncxx::type::k_int32:
+            return static_cast<T>(value.get_int32());
+        case bsoncxx::type::k_double:
+            return static_cast<T>(value.get_double());
+        case bsoncxx::type::k_bool:
+            return static_cast<T>(value.get_bool());
+        default:
+            throw Exception(ErrorCodes::TYPE_MISMATCH, "Type mismatch, {} cannot be converted to number for column {}",
+                            bsoncxx::to_string(value.type()), name);
+    }
+}
 
-                std::vector<std::pair<const Poco::MongoDB::Element *, size_t>> arrays;
-                arrays.emplace_back(&value, 0);
+Array BSONArrayAsArray(size_t dimensions, const bsoncxx::types::b_array & array, const DataTypePtr & type, const Field & default_value, const std::string & name)
+{
+    auto arr = Array();
+    if (dimensions > 0)
```

Comment:
Should we add some santity check for the number of `dimensions` ? Probably ClickHouse won't allow to define too nested array, maybe just in case

Reply:
It's not responsibility of source, is it?
In source we just check data from Mongo, and make sure this can be inserted into given schema.